lib/semantics/check-select-stmt.cc

```diff
@@ -0,0 +1,121 @@
+  // R1149 select-rank-stmt checks
+  std::visit(
+      common::visitors{
+          [](const parser::Expr &) {
+            // TODO:1.Checks if selector is an
+            // expression ex.Rank(a+b)
+          },
+          [&](const parser::Variable &variable) {
+            if (const Symbol * entity{GetLastName(variable).symbol}) {
+              if (!IsAssumedRankArray(*entity)) {  // C1150
+                context_.Say(variable.GetSource(),
+                    "Selector is not an assumed-rank array variable"_err_en_US);
+              }
+            }
+          },
+      },
+      selectRankStmtSel.u);
+  // R1150 select-rank-case-stmt checks
+  auto &rankCaseList{std::get<std::list<parser::SelectRankConstruct::RankCase>>(
+      selectRankConstruct.t)};
+  bool defaultRankFound{false};
+  bool starRankFound{false};
+  std::set<std::int64_t> matches;
+
+  for (const auto &rankCase : rankCaseList) {
+    const auto &rankCaseStmt{
+        std::get<parser::Statement<parser::SelectRankCaseStmt>>(rankCase.t)};
+    const auto &rank{
+        std::get<parser::SelectRankCaseStmt::Rank>(rankCaseStmt.statement.t)};
+    std::visit(
+        common::visitors{
+            [&](const parser::Default &) {  // C1153
+              if (!defaultRankFound) {
+                defaultRankFound = true;
+              } else {
+                context_.Say(rankCaseStmt.source,
+                    "Not more than one of the selectors of SELECT RANK "
+                    "statement may be default"_err_en_US);
+              }
+            },
+            [&](const parser::Star &) {  // C1153
+              if (!starRankFound) {
+                starRankFound = true;
+              } else {
+                context_.Say(rankCaseStmt.source,
+                    "Not more than one of the selectors of SELECT RANK "
+                    "statement may be '*'"_err_en_US);
+              }
+              std::visit(
+                  common::visitors{
+                      [](const parser::Expr &) {
+                        // TODO:1.Checks if selector is an
+                        // expression ex.Rank(a+b)
+                      },
+                      [&](const parser::Variable &variable) {
+                        // C1155 RANK (*) not allowed if the selector has the
+                        // ALLOCATABLE or POINTER attribute
+                        if (const Symbol *
+                            entity{GetLastName(variable).symbol}) {
+                          if (IsAllocatableOrPointer(*entity)) {
+                            context_.Say(variable.GetSource(),
+                                "RANK (*) cannot be used when selector is "
+                                "POINTER or ALLOCATABLE"_err_en_US);
+                          }
+                        }
+                      },
+                  },
+                  selectRankStmtSel.u);
+            },
+            [&](const parser::ScalarIntConstantExpr &init) {
```

Comment:
Within the scope of the `RANK` case, the *associate-name* in the symbol table needs to have the corresponding rank.  Is the name resolution phase already handling this task?